docs/proposals/local-registries.md

```diff
@@ -0,0 +1,165 @@
+
+### Goals
+
+- Simplify how local image registries can be referenced in `Build` objects.
+- Provide a means of declaring that an image references a cluster-local image registry.
+
+### Non-Goals
+
+- Automate how pull/push secrets can be added to a `Build` object.
+- Support secure image pull/push if a local registry uses self-signed certificates.
+
+## Proposal
+
+Application developers will be able to specify `local: true|false` if an image reference points
+to a cluster-local image registry.
+
+The build controller/operator will watch the global "LocalRegistryHosting" `ConfigMap` to check if
+a local image registry is declared. If declared, it will infer the local registry host information
+from what is provided in the `ConfigMap`.
+
+When a `BuildRun` is instantiated, the image specs that declare `local: true` are mutated to inject
+the local image registry host.
+
+### User Stories [optional]
+
+As an application developer
+I would like to declare that images in builds reference a local image registry
+So that I do not have to provide host information in the pull spec
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Obtaining local registry information
+
+First, clusters which deploy a local image registry must provide the releavant host info by
+publishing it in the `local-registry-hosting`
+[ConfigMap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap).
+The build operator/controller must have RBAC which allows it to read the `ConfigMap` and obtain the
+registry host information. For `Build` and `BuildRun` objects, the `HostFromClusterNetwork` value
+should be used as the hostname.
+
+If the `local-registry-hosting` ConfigMap is created, updated, or deleted, the build controller
+and/or operator needs to react accordingly.
+
+#### Declaring an image as local
+
+`Image` references in the API will be updated to have the `local` field. An example using the 
+`output` spec:
+
+```yaml
+spec:
+  output:
+    image: mynamespace/myapp:latest
+    local: true
+```
+
+When an image is declared as `local`, the build controller prepends the local image registry
+hostname to the provided image reference.
+
+### Risks and Mitigations
+
+App developers use `local: true` on clusters that don't have a local registry defined, end up
+pulling images from `docker.io` or an image registry in container runtime's unqualified search
+path.
+
+Local image registry configurations can be altered while the build controller is running.
+
+## Design Details
+
+### Test Plan
+
+1. Deploy the build operator/controller on a cluster that populates the `local-registry-hosting`
+   ConfigMap OR as cluster-admin, manually populate the `local-registry-hosting` ConfigMap.
+2. Create a `Build` and/or `BuildRun` which references a local image.
+3. Ensure that a build can push an image referencing the local image spec.
+
```

Comment:
I suggest a test that specifies a runtime image. Because then the runtime Dockerfile's `FROM` needs to be adjusted.

Reply:
@SaschaSchwarze0 what do you mean? its just about the image that will be pushed. All other images that might required to be pull during the Build process can still use another registry(_e.g. gcr_).

Reply:
We should also test source-to-image builds which use the `builder` image as input.

Reply:
Sry, had missed @qu1queee's comment. What I mean is the following:

- the user specifies a runtime image in the build and the output image to use local
- the build strategy steps will now be configured to push the image to this local registy
- the `runtime_image.go` file adds two more steps to the taskrun and both have to honor the fact that a local registry is used:

1. The `runtime-dockerfile` step which produces the runtime Dockerfile will need to produce a `FROM` statement that loads includes the host name of the internal registry. https://github.com/redhat-developer/build/blob/3a728a0abedf36f6cb6dad7f2e9fa9b125a651ff/pkg/controller/buildrun/runtime_image.go#L21
2. The `kaniko-build-and-push` step (I just noticed that this step is called the same as in our sample kaniko strategy, I suggest to use a different one) will need to specify the `--destination` argument correctly to point to the local registry. https://github.com/redhat-developer/build/blob/3a728a0abedf36f6cb6dad7f2e9fa9b125a651ff/pkg/controller/buildrun/runtime_image.go#L212